sec1_intro_and_setup/test_app/config/initializers/assets.rb

```diff
@@ -0,0 +1,14 @@
+# Be sure to restart your server when you modify this file.
```

Comment:
Whats this file for?

Reply:
This file allows you to add assets to the load path. This file is related to Asset Pipeline which serves pages in your Rails app with the specific assets (CSS, JS, etc) they need.

Reply:
1. What is asset pipeline?
2. How does asset pipeline work together with webpack?

Reply:
1. Asset pipeline provides a framework that helps the webpage to load faster by concatenating and compressing JS and CSS assets.
2. 
- How: Webpack bundles the code into modules and put them in `app/assets/javascripts` to make them available in the asset pipeline.
- Why need webpack to bundle the code: Although asset pipeline can handle everything by itself without webpack, asset pipeline bundles everything(CSS, JS, etc) and loads everything every time, which is unnecessary for the users if the application is large. A better approach is to bundle only the dependencies needed. Webpack allows you to declare the dependencies needed and build a dependency graph to ease future maintenance. Hence, webpack is usually used to handle projects with massive JS. In conclusion, webpack offers an approach to the asset pipeline with better modularisation and dependency loading by bundling the dependencies needed.

Question:
1. How does webpack not make conflict with the asset pipeline?

Reply:
1. Not 100% accurate. Read more
2. Completely inaccurate. Read up more. once you understand this and how rails use both, it will answer your Q

Reply:
1. Asset pipeline provides a framework for connecting, simplifying, or compressing JS and CSS static asset files. It allows us to use other languages ​​and preprocessors, such as CoffeeScript, Sass, and ERB, to write these static resource files. The static resource files in the application can also be automatically merged with the static resource files in other gems.
2. When it comes to asset bundling, Webpack manages JS and Sprockets manages everything else. (Note that webpack treats everything as a JavaScript module, including CSS and images). However, these two processes are completely separated and they do not share dependencies.

Reply:
How does the asset pipeline handles (1)?

Reply:
- For connecting to static resources, Sprockets connects all JavaScript files into one main  `.js` file and all CSS files into one main  `.css` file. This reduces the number of requests initiated by the browser when rendering web pages.
- For simplifying and compressing assets, spaces and comments are removed for CSS files. For JS files, we can select built-in options or customize the processing method to do so.
- Asset pipeline supports coding assets using higher-level language and converts them into actual static resource files through pre-compilation.